### DISCUSSION TITLE
Help help CUPS - Python - IPP

### DISCUSSION THREAD

**Discussion by sonicoprod** [OpenPrinting/cups#444] [Q&A]
Hi Everyone,
I adapted a Python photobooth project on a Raspberry Pi where I use the CUPS server to print photos on a Canon Selphy CP1200.
I would like to manage printers errors (paper jam, no paper, no ink...) to avoid the app to crash.
I found out how to get the printer errors with their code when they raise but I am stuck with the way to send commands to the CUPS server to flush the jobs queue and to restart the printer.
I dont't understand how I can communicate with the server. I think the IPP protocol would be a good way but I don't know how to do ?  I didn't find many infos about this topic.
I am not a professional coder so I try to find a specialist who could explain to me how it works and may be how to do it.
Thanks a lot !
Nico

**Comment by zdohnal** [comment]
Hi @sonicoprod ,

you can use [pyipp](https://pypi.org/project/pyipp/) for IPP communication (I'm not sure how up-to-date the project is with CUPS IPP API though, because I don't use it) or [pycups](https://github.com/OpenPrinting/pycups) for CUPS API written in Python. I'm maintainer of pycups project, and I haven't found a time to update the project with new CUPS API functions from CUPS 2.x, so I would try pyipp project (or combine both of them).

pycups has html documentation in it, and you can see another IPP documentation (not for Python though, but the API is similar and can give an idea how to move on further) - [CUPS programming manual](https://openprinting.github.io/cups/doc/cupspm.html) and [CUPS implementation of IPP](https://openprinting.github.io/cups/doc/spec-ipp.html) and [IPP implementor's guide](http://ftp.pwg.org/pub/pwg/candidates/cs-ippig20-20150821-5100.19.pdf).
